### ISSUE TITLE
Change Request: Support generic types for `Rule.RuleModule` type

### ISSUE THREAD

**Issue by lumirlumir** [eslint/eslint#19521]
### ESLint version

v9.22.0

### What problem do you want to solve?

Hello,

Currently, the `Rule.RuleModule` type is not generic, which prevents me from extending the `ExtRuleDocs` property with custom properties.

Please take a look at the source code below.

https://github.com/eslint/eslint/blob/main/lib/types/index.d.ts#L554-L562

For example:

```js
// @ts-check

/**
 * @typedef {import("eslint").Rule.RuleModule} RuleModule
 */

/** @type {RuleModule} */
export default {
  meta: {
    docs: {
      'my-custom-property': 'my-custom-property' // ❌This line causes a type error❌
      recommended: true,
    },
  },
}
```

When I try to add `my-custom-property` to `meta.docs`, I get a type error because `ExtRuleDocs` is simply initialized as `{}` with no way to extend its functionality.

I’d like to suggest adding support for generic types in `Rule.RuleModule` to allow for greater flexibility.

### What do you think is the correct solution?

- Before: The `RuleModule` type cannot be extended.

```diff
-   type RuleModule = RuleDefinition<{
        LangOptions: Linter.LanguageOptions,
        Code: SourceCode,
        RuleOptions: any[],
        Visitor: NodeListener,
        Node: ESTree.Node,
        MessageIds: string,
-       ExtRuleDocs: {}
    }>;
```

- After: The `RuleModule` type can be extended using generics, allowing users to define custom types.

```diff
+   type RuleModule<ExtRuleDocs = {}> = RuleDefinition<{
        LangOptions: Linter.LanguageOptions,
        Code: SourceCode,
        RuleOptions: any[],
        Visitor: NodeListener,
        Node: ESTree.Node,
        MessageIds: string,
+       ExtRuleDocs: ExtRuleDocs,
    }>;
```

As a result, We can define a custom `ExtRuleDocs` type like the code below:

```js
/**
 * @typedef {import("eslint").Rule.RuleModule<{ 'my-custom-property'?: string }>} RuleModule
 */
```

Please note that this change does not affect ESLint's current behavior. Looking forward to your comments on this!

### Participation

- [x] I am willing to submit a pull request for this change.

**Comment by fasttime** [comment]
Thanks for the issue @lumirlumir, I can reproduce the error using your code. I think that `RuleModule` was originally designed as a type for built-in rules, although it's now used inconsistently also for generic, possibly non-JS rules. And built-in rules don't contain any additional `meta.docs` properties.

I'm not sure if adding type parameters to `RuleModule` and encouraging users to use that type for their own implementations is the way we want to go. Note that we have already a generic interface for rules, which is `RuleDefinition` from `@eslint/core`. Maybe it would make sense to re-export `RuleDefinition` type from `ESLint`, and provide a mechanism to override single type options (such as `ExtRuleDocs`) while still applying defaults for the other options. For example:

```ts
type MyRuleDefinition = RuleDefinition<{ ExtRuleDocs: "my-custom-property"; }>;
```

Opinions on this @eslint/eslint-team?

**Comment by nzakas** [comment]
We want to encourage people to use `RuleDefinition` from `@eslint/core` for any non-JS language rules.

For JavaScript, I'd much rather create a `JSRuleDefinition` similar to [`JSONRuleDefinition'](https://github.com/eslint/json/blob/d613ddec3c7fd608ad6f258d455c6e692f5c9bbf/src/types.ts#L127-L138) that is exported from the `eslint` package (although, eventually it will live in the `@eslint/js` package). 

We should not be making any further changes to `RuleModule` or encourage its further use outside of ESLint itself.

**Comment by nzakas** [comment]
As part of this, we need to figure out how we want to expose language-specific `RuleDefinition` types from our packages.

The first package to do this was `@eslint/json`, which exports `JSONRuleDefinition` here:

https://github.com/eslint/json/blob/d099c78318f8ca9a426d233717728304418425a1/src/types.ts#L127-L138

`JSONRuleDefinition` is generic and supports templates for the `messageIds` and rule options. The only change we might potentially want to make for more general usage outside of the plugin is to also allow passing in `ExtRuleDocs` to allow people to customize what's in `meta.docs`. It seems like that approach would also work for `MarkdownRuleDefinition` and `JSRuleDefinition`.

What do you think?

**Comment by lumirlumir** [comment]
@nzakas @fasttime  

Thanks for the deep investigation into this problem!  

To get straight to the point, the proposed `JSRuleDefinition`, `JSONRuleDefinition`, and `MarkdownRuleDefinition` align well with the direction I had in mind and seem like a great approach.  

When requesting type support, my main concerns were:  
1. Whether ESLint could provide a universal and consistent interface for the various languages it aims to support.  
2. Whether the expansion of the various `meta` properties, which were previously allowed in pure JavaScript code, would be feasible.  

Based on the explanations you provided and the approved changes in this PR  (https://github.com/eslint/markdown/pull/324), it looks like these concerns are being addressed. The introduction of a common interface for multiple languages and the ability to extend `meta` properties will resolve many of the issues I was facing regarding type support.  

---  

I also read the blog post [ESLint now officially supports linting of JSON and Markdown](https://eslint.org/blog/2024/10/eslint-json-markdown-support/) that was published in the official documentation website.  

I personally believe that this proposal marks a significant step forward in achieving ESLint’s goal of providing universal linting support across multiple languages!

**Comment by fasttime** [comment]
> `JSONRuleDefinition` is generic and supports templates for the `messageIds` and rule options. The only change we might potentially want to make for more general usage outside of the plugin is to also allow passing in `ExtRuleDocs` to allow people to customize what's in `meta.docs`. It seems like that approach would also work for `MarkdownRuleDefinition` and `JSRuleDefinition`.

I think yes, the customizable properties should be `RuleOptions`, `MessageIds` and `ExtRuleDocs`. The other generic properties of `RuleDefinition` are language-specific.

Currently, in `JSONRuleDefinition` the different properties must be specified as separate type params:

```ty
type MyRule = JSONRuleDefinition<[{ someOption?: boolean, anotherOption?: string }], "foo" | "bar">;
```

I would suggest to add the ability to specify each type property as an optional property of a single object parameter, e.g.:

```ty
type MyRule1 = JSONRuleDefinition<{ RuleOptions: [{ someOption?: boolean; anotherOption?: string; }]; }>;

type MyRule2 = JSONRuleDefinition<{ MessageIds: "foo" | "bar"; }>;

type MyRule3 = JSONRuleDefinition<{ ExtRuleDocs: { someProp: string; }; }>;
```

This way, only properties that must be overridden will have to be indicated explicitly, while other properties will retain their default settings.

**Comment by nzakas** [comment]
@fasttime can you explain how your suggestion would work? Can we specify defaults for each property?

**Comment by fasttime** [comment]
There are probably different ways to specify defaults for properties of type parameters. What I tried in `eslint/markdown` is using a combination of optional properties and their default types with `Required<Options & Omit<DefaultOptions, keyof Options>>`, where `Omit<DefaultOptions, keyof Options>` provides defaults for properties not included in `Options`. `Required<...>` is used to exclude `undefined` types that TypeScript implicitly assumes as valid for optional properties.

For JavaScript rules in this repo, the type would look roughly like this:

```ts
type JsRuleDefinitionTypeOptions = {
	RuleOptions: unknown[];
	MessageIds: string;
	ExtRuleDocs: Record<string, unknown>;
};

type JsRuleDefinition<Options extends Partial<JsRuleDefinitionTypeOptions> = {}> =
RuleDefinition<

	// JavaScript-specific properties
	{
		LangOptions: Linter.LanguageOptions;
		Code: SourceCode;
		Visitor: NodeListener;
		Node: ESTree.Node;
	} & Required<

	// Rule-specific properties
	Options &

	// Defaults for rule-specific properties
	Omit<JsRuleDefinitionTypeOptions, keyof Options>

>>;
```

The above doesn't include a fallback for the case where each option is passed as a separate type parameter, as it is currently done in `eslint/json`, for example:

```ts
type MyRule = JSONRuleDefinition<[{ myOption?: number; }], "message1" | "message2">;
```

This fallback could be achieved with a helper type. Not completely tested, but should be more or less like this:

```ts
type JsRuleDefinition<Options extends Partial<JsRuleDefinitionTypeOptions> | unknown[] = {}, MessageIds = string> =
JsRuleDefinitionHelper<
	Options extends unknown[] ?
	{ RuleOptions: Options; MessageIds: MessageIds extends string ? MessageIds : never; } :
	Options
>;
```

I can prepare a prototype if we want to try this approach.

**Comment by fasttime** [comment]
It looks like we could export `JSONRuleDefinition` from `@eslint/json` and `MarkdownRuleDefinition` from `@eslint/markdown`, and add a similar generic type for `JsRuleDefinition` in this repo?

**Comment by nzakas** [comment]
@fasttime yup, that's what I'm thinking, though we should do `JSRuleDefinition` as we've used all-caps for most abbreviations.

**Comment by fasttime** [comment]
`JSRuleDefinition` sounds good, marking as accepted.

@lumirlumir would you still be interested in submitting a pull request to add a `JSRuleDefinition` generic type in this repo as per the above discussion?

**Comment by lumirlumir** [comment]
I'd love to if I can!

I'll follow the current implementation of [JSON](https://github.com/eslint/json/blob/main/src/types.ts#L71-L92) and [Markdown](https://github.com/eslint/markdown/blob/main/src/types.ts#L95-L116).

**Comment by fasttime** [comment]
> [@fasttime](https://github.com/fasttime) I'd love to if I can!
> 
> I'll follow the current implementation of [JSON](https://github.com/eslint/json/blob/main/src/types.ts#L71-L92) and [Markdown](https://github.com/eslint/markdown/blob/main/src/types.ts#L95-L116).

Thanks! A pull request is welcome.
